### ISSUE TITLE
Remember me in combination with Kerberos and 2FA

### ISSUE THREAD

**Issue by joeri-jansen** [keycloak/keycloak#22546]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

authentication

### Describe the bug

We have a configuration that allows users to login via kerberos (80% of the users) or username/password. Second, we require 2FA (webauthn or OTP). The realm supports 'Remember Me' to reduce the amount of requests for username / password and 2FA. 

For users who do not login via Kerbers, this flow works fine: using the remember me option in the username / password form, and entering the 2FA, the users do not have to re-authenticate if they close / reopen the browser within the idle session timeout, because the cookie is persistent. This is the desired flow: we require 1 OTP/webauthn challenge a day.

However, users with Kerberos authentication do not get the option to select 'remember me' on the username password form, but do get a Webauth / OTP prompt. If these users close and reopen the browser, they get the Webauthn/OTP prompt again: there is no option to make the session persistent.
To resolve this, we added the 'remember me' option to the OTP and webauthn forms in our customized theme. However, this option does not seem to be effective / handled upon submission of the OTP and webauthn form: even with the checkbox filled, the cookie stays 'browser session only'. This results in users being prompted for a 2FA multiple times a day.

### Version

22.0.1

### Expected behavior

Authentication flows without username / password prompt, with 2FA enabled, should support the 'remember me' function to reduce 2FA prompts upon browser restart.
The web authentication handling accepts the rememberMe form variable in other steps, like submission of OTP or webauthn

### Actual behavior

The web authentication handling ignores the rememberMe form variable in case of a OTP or webauthn response.

### How to Reproduce?

We have an browser authentication flow that uses:

- cookie (alternative)
- login flow + 2FA (alternative)
- - first factor flow (required)
- - - kerberos (alternative)
- - - username password form (alternative)
- - second factor (required)
- - - WebAutn (alternative)
- - - OTP (alternative)

The realm supports 'Remember Me'. Kerberos is configured.

Customize the theme to add a remember me option to the login-otp.ftl

**Comment by rmartinc** [comment]
@joeri-jansen The remember-me is only implemented in the username password form. If you see the [AbstractUsernameFormAuthenticator](https://github.com/keycloak/keycloak/blob/22.0.3/services/src/main/java/org/keycloak/authentication/authenticators/browser/AbstractUsernameFormAuthenticator.java#L209-L217) it adds the note to the session. The OTP or Spnego don't do that and they never create a remember-me session. So I think this is expected behavior.

But, anyway, IIRC the remember me just works to set the user to login (the username field in the textbox is filled with the user) and also uses some other timeouts. What are the timeouts you are setting in the realm (Realm Settings -> Sessions -> SSO Session Idle, SSO Session Max, SSO Session Idle Remember Me and SSO Session Max Remember Me)? Are they different for SSO and Remember Me? That can explain the difference too.

**Comment by joeri-jansen** [comment]
@rmartinc the main goal of the change is to keep the session, even on browser restarts. We currently use a the following timeouts:
- SSO session idle: 10 hours
- SSO session max: 10 hours
- SSO session idle remember me: not set (10 hours effective)
- SSO session max remember me: not set (10 hours effective)

Since we use Kerberos for user name and password, the user only enters the OTP. However, even within the window of 10 hours, upon browser restart, the user must enter a new OTP. That is the main problem.

We now extended the Keycloak SpnegoAuthenticator, to set 'rememberme' upon succesfull Kerberos authentication:
`@Override
    public void authenticate(AuthenticationFlowContext context) {
		super.authenticate(context);
		if (context.getUser() != null) {
			// Kerberos authentication successful, user is logged in with own AD-account
			if (context.getRealm().isRememberMe()) { 
				// Make session persistent if realm allows it
				context.getAuthenticationSession().setAuthNote(Details.REMEMBER_ME, "true");
	            context.getEvent().detail(Details.REMEMBER_ME, "true");
			}
		}
	}`
  
This solves the OTP-request upon browser restart within the SSO session window. Since the user is authenticated with domain credentials in it's own user profile, there is no risk of keeping the session 'remember me'.

Perhaps a nice feature if this could be made configurable on the existing SpnegoAuthenticator. I could create a pull request if other users are interested...

**Comment by rmartinc** [comment]
I checked this a bit more and I realized that the `KEYCLOAK_IDENTITY` cookie is created with max-age -1 if no remember-me is set ([here](https://github.com/keycloak/keycloak/blob/22.0.3/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java#L785-L789)). That means that the cookie is deleted when the browser is closed. But that doesn't happen if your browser is configured to restore tabs, and that's why I didn't even notice it, I had my browser set this way.

Let me think about this. Because I'm not sure if we should always add this just if the remember-me is set at realm level.

**Comment by rmartinc** [comment]
I've been thinking about this, and maybe we need to add a config property in the kerberos authenticator that enables the remember me (the option would be defaulted to false if not found). This way, if the realm is configured to use remember-me and the spnego authenticator is configured to use remember-me the code to force the remember-me in the session is executed. We are not modifying current behavior but the remember-me option can be configured. The config property can be extended to more authenticators if needed. It would be something similar to what is used for example in the [IdentityProviderAuthenticatorFactory](https://github.com/keycloak/keycloak/blob/22.0.3/services/src/main/java/org/keycloak/authentication/authenticators/browser/IdentityProviderAuthenticatorFactory.java#L76).

@joeri-jansen @mposolda WDYT?

This is then an enhancement and I'm moving to backlog. PRs are welcomed as always.

**Comment by joeri-jansen** [comment]
@rmartinc I was also thinking of making it configurable for the Kerberos authenticator. Another, perhaps even better solution, is to create an additional authenticator that sets this flag during the browser login flow (it should be added after the login and 2FA steps and require an authenticated user). 

That would also solve this problem in loginflows where the primary authentication is delegated to an external Idp (no option to keep the Keycloak session alive) and Keycloak adds a 2FA after successful external authentication. That is a flow I also applied for several customers. Also, the addition of a small new authenticator has the advantage that existing flows don't change behaviour.

**Comment by rmartinc** [comment]
@joeri-jansen Yes, it's a good idea too, that way you just need to modify the flow adding the new authenticator to setup the remember-me. It works for me.

I'm removing myself from the issue. If you want to work on it just let me know.
